{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Pre-Process Data: Specific DSAID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Pre-Process Data\n",
    "\n",
    "Convert the raw data counts into sc-RNAseq compatible data format.\n",
    "\n",
    "Structure:\n",
    "    1. Imports, Variables, Functions\n",
    "    2. Load Data\n",
    "    3. Convert to `adata` object\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "# 1. Imports, Variables, Functions\n",
    "# imports\n",
    "import numpy as np, os, sys, pandas as pd, scanpy as sc\n",
    "import anndata as ad\n",
    "\n",
    "# variables\n",
    "example_data_path = (\n",
    "    \"/aloy/home/ddalton/projects/disease_signatures/data/DiSignAtlas/tmp/DSA00015.csv\"\n",
    ")\n",
    "\n",
    "# functions\n",
    "\n",
    "\n",
    "# 2. Load Data\n",
    "df = pd.read_csv(example_data_path, index_col=0)\n",
    "\n",
    "# 3. Convert to `adata` object\n",
    "\n",
    "\n",
    "# Extract cell identifiers and gene expression data\n",
    "cell_ids = df.iloc[:, 0]\n",
    "gene_expression_data = df.iloc[:, 1:].values\n",
    "gene_names = df.columns[1:]\n",
    "\n",
    "\n",
    "# Create an AnnData object\n",
    "adata = ad.AnnData(X=gene_expression_data)\n",
    "\n",
    "\n",
    "# Add cell and gene metadata\n",
    "adata.obs[\"cell_ids\"] = cell_ids.values\n",
    "adata.var[\"gene_symbols\"] = gene_names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Preprocess Data: Specific Diseases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df_info' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 77\u001b[0m\n\u001b[1;32m     73\u001b[0m             df_global \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat([df_global, __df], axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m)\n\u001b[1;32m     74\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m df_global\n\u001b[0;32m---> 77\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mget_dataset_info\u001b[39m(ids, df_info\u001b[38;5;241m=\u001b[39m\u001b[43mdf_info\u001b[49m):\n\u001b[1;32m     78\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Get Disease Information\"\"\"\u001b[39;00m\n\u001b[1;32m     79\u001b[0m     dsaid_2_accession \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mdict\u001b[39m(\u001b[38;5;28mzip\u001b[39m(df_info[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdsaid\u001b[39m\u001b[38;5;124m\"\u001b[39m], df_info[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124maccession\u001b[39m\u001b[38;5;124m\"\u001b[39m]))\n",
      "\u001b[0;31mNameError\u001b[0m: name 'df_info' is not defined"
     ]
    }
   ],
   "source": [
    "\"\"\"Pre-Process Data\n",
    "\n",
    "Convert the raw data counts into sc-RNAseq compatible data format.\n",
    "\n",
    "Structure:\n",
    "    1. Imports, Variables, Functions\n",
    "    2. Load Data\n",
    "    3. Convert to `adata` object\n",
    "\n",
    "\"\"\"\n",
    "\n",
    "# 1. Imports, Variables, Functions\n",
    "# imports\n",
    "import numpy as np, os, sys, pandas as pd, scanpy as sc\n",
    "import anndata as ad\n",
    "import logging\n",
    "from tqdm import tqdm\n",
    "\n",
    "logging.basicConfig(level=logging.INFO, format=\"%(asctime)s - %(message)s\")\n",
    "from matplotlib import pyplot as plt\n",
    "from collections import Counter\n",
    "\n",
    "# import random\n",
    "import random\n",
    "\n",
    "from sklearn.model_selection import train_test_split\n",
    "\n",
    "\n",
    "# variables\n",
    "diseases_of_interest_set = {\"Influenza\", \"Colorectal Carcinoma\", \"Breast Cancer\"}\n",
    "diseases_of_interest_set = {\n",
    "    \"Crohn's Disease\",\n",
    "    \"Ulcerative Colitis\",\n",
    "    \"Lung Cancer\",\n",
    "    \"Lung Adenocarcinoma\",\n",
    "    \"Breast Cancer\",\n",
    "    \"Psoriasis\",\n",
    "}\n",
    "\n",
    "\n",
    "example_data_path = (\n",
    "    \"/aloy/home/ddalton/projects/disease_signatures/data/DiSignAtlas/tmp/DSA00123.csv\"\n",
    ")\n",
    "\n",
    "df_info_path = os.path.join(\n",
    "    \"/aloy\",\n",
    "    \"home\",\n",
    "    \"ddalton\",\n",
    "    \"projects\",\n",
    "    \"disease_signatures\",\n",
    "    \"data\",\n",
    "    \"DiSignAtlas\",\n",
    "    \"Disease_information_Datasets_extended.csv\",\n",
    ")\n",
    "\n",
    "\n",
    "large_df_path = \"/aloy/home/ddalton/projects/disease_signatures/data/DiSignAtlas/DiSignAtlas.exp_prof_merged.csv\"\n",
    "\n",
    "\n",
    "# functions\n",
    "def get_exp_prof(dsaids_interest):\n",
    "    \"\"\"Get Expression Profiles\"\"\"\n",
    "\n",
    "    # variables\n",
    "    file_dir = \"/aloy/home/ddalton/projects/disease_signatures/data/DiSignAtlas/tmp/\"\n",
    "    first = True\n",
    "    for dsaid in tqdm(dsaids_interest):\n",
    "        __df = pd.read_csv(os.path.join(file_dir, f\"{dsaid}.csv\"))\n",
    "        if first:\n",
    "            df_global = __df\n",
    "            first = False\n",
    "        else:\n",
    "            df_global = pd.concat([df_global, __df], axis=0)\n",
    "    return df_global\n",
    "\n",
    "\n",
    "def get_dataset_info(ids, df_info):\n",
    "    \"\"\"Get Disease Information\"\"\"\n",
    "    dsaid_2_accession = dict(zip(df_info[\"dsaid\"], df_info[\"accession\"]))\n",
    "\n",
    "    dataset_accessions = [dsaid_2_accession[id.split(\";\")[0]] for id in ids]\n",
    "\n",
    "    accession_2_id = {k: v for v, k in enumerate(set(dataset_accessions))}\n",
    "    dataset_ids = [accession_2_id[accession] for accession in dataset_accessions]\n",
    "\n",
    "    return dataset_accessions, dataset_ids\n",
    "\n",
    "\n",
    "def get_disease_info(ids, df_info):\n",
    "    \"\"\"Get Disease Information\"\"\"\n",
    "    dsaid_2_disease = dict(zip(df_info[\"dsaid\"], df_info[\"disease\"]))\n",
    "    disease_types = list()\n",
    "    for id in ids:\n",
    "        if id.split(\";\")[2] == \"Control\":\n",
    "            disease_types.append(\"Control\")\n",
    "        elif id.split(\";\")[2] == \"Case\":\n",
    "            disease_types.append(dsaid_2_disease[id.split(\";\")[0]])\n",
    "        else:\n",
    "            print(f\"Error with ID: {id}\")\n",
    "\n",
    "    map_2_id = {k: v for v, k in enumerate(set(disease_types))}\n",
    "\n",
    "    disease_ids = [map_2_id[x] for x in disease_types]\n",
    "\n",
    "    return disease_types, disease_ids\n",
    "\n",
    "\n",
    "# 2. Load Data\n",
    "# df = pd.read_csv(example_data_path)\n",
    "\n",
    "df_info = pd.read_csv(df_info_path)\n",
    "\n",
    "\n",
    "# Query data to retrieve dsaids of interest\n",
    "library_strategies_of_interest_set = {\"RNA-seq\", \"Microarray\"}\n",
    "QUERY = \"disease in @diseases_of_interest_set & library_strategy in @library_strategies_of_interest_set & organism == 'Homo sapiens'\"\n",
    "dsaids_interest = df_info.query(QUERY)[\"dsaid\"].to_list()\n",
    "logging.info(f\"NÂº of DSAIDs of interest: {len(dsaids_interest)}\")\n",
    "\n",
    "# skip_rows_idxs = get_skip_rows(dsaids_interest)\n",
    "\n",
    "df = get_exp_prof(dsaids_interest)\n",
    "\n",
    "# 3. Convert to `adata` object\n",
    "# load dataframe\n",
    "# df = pd.read_csv(large_df_path, skiprows=skip_rows_idxs, index_col=0)\n",
    "logging.info(f\"Loaded dataframe with shape: {df.shape}\")\n",
    "\n",
    "# drop non significant rows\n",
    "# Calculate the number of NaNs in each row\n",
    "nan_counts = df.isna().sum(axis=1)\n",
    "\n",
    "# Filter the DataFrame to keep only rows with NaNs less than or equal to 18,000\n",
    "df = df[nan_counts <= 18000]\n",
    "logging.info(f\"Filtered dataframe with shape: {df.shape}\")\n",
    "\n",
    "# Filter out Unknown samples\n",
    "mask = [False if id.split(\";\")[2] == \"Unknown\" else True for id in df.iloc[:, 0].values]\n",
    "\n",
    "df = df[mask]\n",
    "\n",
    "logging.info(f\"Filtered out Unkowns from dataframe with shape: {df.shape}\")\n",
    "\n",
    "\n",
    "# 3. Convert to `adata` object\n",
    "# Extract cell identifiers and gene expression data\n",
    "ids = df.iloc[:, 0]\n",
    "gene_expression_data = df.iloc[:, 1:].values\n",
    "gene_names = df.columns[1:]\n",
    "\n",
    "# Create an AnnData object\n",
    "adata = ad.AnnData(X=gene_expression_data)\n",
    "\n",
    "\n",
    "# Add cell and gene metadata\n",
    "adata.obs[\"ids\"] = ids.values\n",
    "adata.var[\"gene_symbols\"] = gene_names\n",
    "adata.var[\"index\"] = gene_names\n",
    "adata.var[\"gene_name\"] = gene_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-20 16:41:24,867 - Counts of each disease type: Counter({'Control': 1764, 'Breast Cancer': 1740, 'Colorectal Carcinoma': 1115, 'Influenza': 471})\n"
     ]
    }
   ],
   "source": [
    "# generate celltype and celltype_id columns in adata.obs\n",
    "disease_types, disease_ids = get_disease_info(ids.values, df_info=df_info)\n",
    "adata.obs[\"celltype\"] = disease_types\n",
    "adata.obs[\"celltype_id\"] = disease_ids\n",
    "\n",
    "logging.info(f\"Counts of each disease type: {Counter(disease_types)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-21 09:17:33,808 - Counts of each dataset id: Counter({'GSE54002': 431, 'GSE87211': 361, 'GSE41258': 321, 'GSE55276': 234, 'GSE55277': 234, 'GSE5364': 194, 'GSE45827': 162, 'GSE101702': 157, 'GSE27562': 146, 'GSE24124': 135, 'GSE47756': 127, 'GSE16443': 119, 'GSE42568': 119, 'GSE164191': 119, 'GSE29044': 107, 'GSE37751': 106, 'GSE33113': 94, 'GSE123086': 92, 'GSE9348': 80, 'GSE62932': 79, 'GSE29366': 67, 'GSE29431': 64, 'GSE61304': 58, 'GSE74604': 58, 'GSE10810': 56, 'GSE110933': 48, 'GSE34205': 48, 'GSE36295': 48, 'GSE24514': 47, 'GSE3744': 45, 'GSE139038': 45, 'GSE20437': 40, 'GSE100179': 38, 'GSE100150': 37, 'GSE32488': 37, 'GSE123087': 36, 'GSE3964': 35, 'GSE64392': 35, 'GSE32323': 34, 'GSE22598': 34, 'GSE77953': 30, 'GSE31471': 30, 'GSE80759': 30, 'GSE122183': 30, 'GSE10715': 30, 'GSE9574': 29, 'GSE41011': 29, 'GSE70468': 28, 'GSE14297': 25, 'GSE4183': 23, 'GSE4107': 22, 'GSE27131': 21, 'E-MTAB-5385': 21, 'GSE20266': 20, 'GSE11545': 20, 'GSE90708': 19, 'GSE25487': 19, 'GSE71053': 18, 'GSE33126': 18, 'GSE66387': 18, 'GSE31469': 18, 'GSE71363': 18, 'GSE37911': 18, 'GSE185645': 16, 'GSE36765': 14, 'GSE10714': 13, 'GSE26910': 12, 'GSE157737': 12, 'GSE146587': 12, 'GSE152508': 11, 'GSE28000': 11, 'GSE68849': 10, 'GSE7700': 10, 'GSE18462': 10, 'GSE77199': 8, 'GSE19295': 8, 'GSE96597': 8, 'GSE26990': 8, 'GSE137511': 8, 'GSE65517': 7, 'GSE141174': 6, 'GSE31475': 6, 'GSE31474': 6, 'GSE31470': 6, 'GSE128610': 6, 'GSE83353': 6, 'GSE31138': 6, 'GSE22242': 5, 'GSE73613': 4})\n"
     ]
    }
   ],
   "source": [
    "# generate dataset ids column in adata.obs\n",
    "dataset_ids, batch_ids = get_dataset_info(ids.values, df_info=df_info)\n",
    "\n",
    "adata.obs[\"dataset_id\"] = dataset_ids\n",
    "adata.obs[\"batch_id\"] = batch_ids\n",
    "\n",
    "logging.info(f\"Counts of each dataset id: {Counter(dataset_ids)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2024-07-20 17:19:19,651 - NÂº of single label indexes: 2\n",
      "2024-07-20 17:19:19,656 - NÂº of train samples: 4072\n",
      "2024-07-20 17:19:19,657 - NÂº of test samples: 1018\n"
     ]
    }
   ],
   "source": [
    "# We combine both the disease type with the dataset as to\n",
    "# shuffle even more the data - minimize bias in train/test split\n",
    "labels = np.array([a + b for a, b in zip(disease_types, dataset_ids)])\n",
    "\n",
    "# Generate indices for the data points\n",
    "indices = np.arange(len(labels))\n",
    "\n",
    "\"\"\"train_test_split cannot handle single label indexes\n",
    "\n",
    "Because of this we will manually deal with these cases!\n",
    "\n",
    "\"\"\"\n",
    "indices_single_label = [i for i, x in enumerate(labels) if list(labels).count(x) == 1]\n",
    "labels_single_label = labels[indices_single_label]\n",
    "\n",
    "logging.info(f\"NÂº of single label indexes: {len(indices_single_label)}\")\n",
    "\n",
    "\n",
    "remaining_indices = [i for i in indices if i not in indices_single_label]\n",
    "remaining_labels = labels[remaining_indices]\n",
    "\n",
    "# Perform stratified split on the remaining indices\n",
    "train_indices, test_indices = train_test_split(\n",
    "    remaining_indices, test_size=0.2, stratify=remaining_labels, random_state=42\n",
    ")\n",
    "\n",
    "for idx in indices_single_label:\n",
    "\n",
    "    if random.random() < 0.2:\n",
    "        test_indices.append(idx)\n",
    "    else:\n",
    "        train_indices.append(idx)\n",
    "\n",
    "\n",
    "test_indices.sort()\n",
    "train_indices.sort()\n",
    "\n",
    "batch_ids = np.zeros(len(labels))\n",
    "batch_ids[test_indices] = 1\n",
    "\n",
    "\n",
    "logging.info(f\"NÂº of train samples: {np.sum(~batch_ids.astype(bool))}\")\n",
    "logging.info(f\"NÂº of test samples: {np.sum(batch_ids.astype(bool))}\")\n",
    "\n",
    "\n",
    "adata.obs[\"train_test\"] = batch_ids\n",
    "adata.obs[\"str_batch\"] = batch_ids.astype(int).astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'str_batch' as categorical\n"
     ]
    }
   ],
   "source": [
    "# save adata object\n",
    "# adata.write(\"../data/test_1/test_1.h5ad\")\n",
    "adata.write(\"../data/test_1/test_2.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ids</th>\n",
       "      <th>celltype</th>\n",
       "      <th>celltype_id</th>\n",
       "      <th>dataset_id</th>\n",
       "      <th>batch_id</th>\n",
       "      <th>train_test</th>\n",
       "      <th>batch_str</th>\n",
       "      <th>str_batch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSA00057;GSM2062351;Control</td>\n",
       "      <td>Control</td>\n",
       "      <td>0</td>\n",
       "      <td>GSE77953</td>\n",
       "      <td>23</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DSA00057;GSM2062352;Control</td>\n",
       "      <td>Control</td>\n",
       "      <td>0</td>\n",
       "      <td>GSE77953</td>\n",
       "      <td>23</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSA00057;GSM2062353;Control</td>\n",
       "      <td>Control</td>\n",
       "      <td>0</td>\n",
       "      <td>GSE77953</td>\n",
       "      <td>23</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DSA00057;GSM2062354;Control</td>\n",
       "      <td>Control</td>\n",
       "      <td>0</td>\n",
       "      <td>GSE77953</td>\n",
       "      <td>23</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSA00057;GSM2062355;Control</td>\n",
       "      <td>Control</td>\n",
       "      <td>0</td>\n",
       "      <td>GSE77953</td>\n",
       "      <td>23</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5085</th>\n",
       "      <td>DSA10295;GSM1012438;Case</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>3</td>\n",
       "      <td>GSE41258</td>\n",
       "      <td>16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5086</th>\n",
       "      <td>DSA10295;GSM1012444;Case</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>3</td>\n",
       "      <td>GSE41258</td>\n",
       "      <td>16</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5087</th>\n",
       "      <td>DSA10295;GSM1012448;Case</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>3</td>\n",
       "      <td>GSE41258</td>\n",
       "      <td>16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5088</th>\n",
       "      <td>DSA10295;GSM1012449;Case</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>3</td>\n",
       "      <td>GSE41258</td>\n",
       "      <td>16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5089</th>\n",
       "      <td>DSA10295;GSM1012450;Case</td>\n",
       "      <td>Colorectal Carcinoma</td>\n",
       "      <td>3</td>\n",
       "      <td>GSE41258</td>\n",
       "      <td>16</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5090 rows Ã 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              ids              celltype  celltype_id  \\\n",
       "0     DSA00057;GSM2062351;Control               Control            0   \n",
       "1     DSA00057;GSM2062352;Control               Control            0   \n",
       "2     DSA00057;GSM2062353;Control               Control            0   \n",
       "3     DSA00057;GSM2062354;Control               Control            0   \n",
       "4     DSA00057;GSM2062355;Control               Control            0   \n",
       "...                           ...                   ...          ...   \n",
       "5085     DSA10295;GSM1012438;Case  Colorectal Carcinoma            3   \n",
       "5086     DSA10295;GSM1012444;Case  Colorectal Carcinoma            3   \n",
       "5087     DSA10295;GSM1012448;Case  Colorectal Carcinoma            3   \n",
       "5088     DSA10295;GSM1012449;Case  Colorectal Carcinoma            3   \n",
       "5089     DSA10295;GSM1012450;Case  Colorectal Carcinoma            3   \n",
       "\n",
       "     dataset_id  batch_id  train_test  batch_str  str_batch  \n",
       "0      GSE77953        23         0.0        0.0        0.0  \n",
       "1      GSE77953        23         1.0        1.0        1.0  \n",
       "2      GSE77953        23         0.0        0.0        0.0  \n",
       "3      GSE77953        23         0.0        0.0        0.0  \n",
       "4      GSE77953        23         1.0        1.0        1.0  \n",
       "...         ...       ...         ...        ...        ...  \n",
       "5085   GSE41258        16         0.0        0.0        0.0  \n",
       "5086   GSE41258        16         1.0        1.0        1.0  \n",
       "5087   GSE41258        16         0.0        0.0        0.0  \n",
       "5088   GSE41258        16         0.0        0.0        0.0  \n",
       "5089   GSE41258        16         0.0        0.0        0.0  \n",
       "\n",
       "[5090 rows x 8 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# we will create an \"artifical\" cell type which\n",
    "# will represent the diseases"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scgpt",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
